### PR TITLE
resolve name to ip before passing to csi-proxy

### LINKED ISSUE #294: smb-server.default.svc.cluster.local could not be recognized by New-SmbGlobalMapping command on Windows node
**What happened**:
FailedMount  3s (x6 over 29s)  kubelet            MountVolume.MountDevice failed for volume "pvc-34f4bd50-0b59-4118-b814-13c3b024a478" : rpc error: code = Internal desc = volume(pvc-34f4bd50-0b59-4118-b814-13c3b024a478) mount "\\\\smb-server.default.svc.cluster.local\\share" on "\\var\\lib\\kubelet\\plugins\\kubernetes.io\\csi\\pv\\pvc-34f4bd50-0b59-4118-b814-13c3b024a478\\globalmount" failed with smb mapping failed with error: rpc error: code = Unknown desc = NewSmbGlobalMapping failed. output: "New-SmbGlobalMapping : The network path was not found. \r\nAt line:1 char:190\r\n+ ... ser, $PWord;New-SmbGlobalMapping -RemotePath $Env:smbremotepath -Cred ...\r\n+                 ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~\r\n    + CategoryInfo          : NotSpecified: (MSFT_SmbGlobalMapping:ROOT/Microsoft/...mbGlobalMapping) [New-SmbGlobalMa \r\n   pping], CimException\r\n    + FullyQualifiedErrorId : Windows System Error 53,New-SmbGlobalMapping\r\n \r\n", err: exit status 1

**What you expected to happen**:
Mount succeed!

**How to reproduce it**:
1. Start a cluster with Windows nodes
2. Install CSI driver
3. Install a samba server 
4. Deploy a storage class
    kubectl apply -f https://gist.githubusercontent.com/lizhuqi/087c1cf8410f1a9ba681ee4be25118c9/raw/5d434991feef1b4da77aaa87501722b7c7e0e1fa/storageclass-smb-name.yaml
5. Deploy a stateful set
     kubectl apply -f https://gist.githubusercontent.com/lizhuqi/be85a31055e8f0dfc94e8c94c73b33dd/raw/ed56a5e0af452a2ca3dcda00b5ce261cc601b8cf/statefulset-win-name.yaml

**Anything else we need to know?**:
Using public IP or cluster IP or SMB server name in a AD domain works but smb-server.default.svc.cluster.local for the samba server doesn't work. I've confirmed DNS for the smb-server.default.svc.cluster.local works fine.

kubectl exec [POD_NAME] -- nslookup smb-server.default.svc.cluster.local
Server:         10.20.0.10
Address:        10.20.0.10#53

Name:   smb-server.default.svc.cluster.local
Address: 10.20.10.72

From the error, it seems that New-SmbGlobalMapping failed to recognize smb-server.default.svc.cluster.local.

**Environment**:
- CSI Driver version: latest
- Kubernetes version (use `kubectl version`): 1.20.6

### PR DESCRIPTION
**Pull request by lizhuqi**
**What type of PR is this?**
/kind feature

**What this PR does / why we need it**:
csi-proxy is running in hostnetwork so it can't resolve hostname of k8s services to IP. This PR resolves name to ip before passing to csi-proxy.

**Which issue(s) this PR fixes**:
Fixes #294

**Special notes for your reviewer**:

**Release note**:
```
none
```

### CODE REVIEW

**Review by andyzhangx — COMMENTED** [review]
/ok-to-test

File: pkg/mounter/safe_mounter_windows.go [hunk 1]
```diff
@@ -71,6 +72,17 @@ func (mounter *CSIProxyMounter) SMBMount(source, target, fsType string, mountOpt
 		}
 	}
 
+	parts := strings.FieldsFunc(source, Split)
+	if len(parts) > 0 && strings.HasSuffix(parts[0], "default.svc.cluster.local") {
```
Comment (andyzhangx) [inline pkg/mounter/safe_mounter_windows.go:76]: `svc.cluster.local`

File: pkg/mounter/safe_mounter_windows.go [hunk 2]
```diff
@@ -71,6 +72,17 @@ func (mounter *CSIProxyMounter) SMBMount(source, target, fsType string, mountOpt
...
+		// replace hostname with IP in the source
+		ip, err := net.ResolveIPAddr("ip4", parts[0])
+		if err != nil {
```
Comment (andyzhangx) [inline pkg/mounter/safe_mounter_windows.go:79]: if there is error, log a warning msg here, and then continue.
And no matter it resolve IP succeed or not, should always log here.

File: pkg/mounter/safe_mounter_windows.go [hunk 3]
```diff
@@ -71,6 +72,17 @@ func (mounter *CSIProxyMounter) SMBMount(source, target, fsType string, mountOpt
 		}
 	}
 
+	parts := strings.FieldsFunc(source, Split)
+	if len(parts) > 0 && strings.HasSuffix(parts[0], "default.svc.cluster.local") {
+		// replace hostname with IP in the source
+		ip, err := net.ResolveIPAddr("ip4", parts[0])
```
Comment (andyzhangx) [inline pkg/mounter/safe_mounter_windows.go:78]: use a new var here:
```
domainName := parts[0]
```

File: pkg/mounter/safe_mounter_windows.go (hunk 2 above)
Comment (lizhuqi) [reply pkg/mounter/safe_mounter_windows.go:80]: Updated. Please take another look.

**Review by andyzhangx — APPROVED** [review]
/lgtm
/approve

**Pull request merged**